plaidml/bridge/openvino/ops/deformable_convolution.cpp

```diff
@@ -0,0 +1,605 @@
+      for (auto off_w = 0; off_w < OFF_W; ++off_w) {
+        for (auto f_w = 0; f_w < F_W; ++f_w) {
+          auto data = off_w * strides[0] + f_w * dilations[0] - pad_befores[0];
+          data_w.push_back(static_cast<float>(data));
+        }
+      }
+      buffer_w.copy_from(data_w.data());
+      auto input_index_w = Constant(buffer_w, "input_index_w");
+      input_index_w = op::reshape(input_index_w, make_tuple<int64_t>({1, 1, NEW_W}));
+      // Get deformabled index.
+      edsl::Tensor new_index_w = offset_w + input_index_w;
+      // Get deformable input tensor.
+      edsl::Tensor deform_input = edsl::gather(I, new_index_w).axis(-1);
+      deform_input = extract_tensor(deform_input, rank - 2);
+      // Compute DeformableConvolution.
+      autopad_mode = plaidml::op::AutoPadMode::VALID;
+      auto result = op::convolution(deform_input, F)
+                        .strides(std::vector<size_t>({(size_t)F_W}))
+                        .autopad_mode(autopad_mode)
+                        .input_layout(plaidml::op::TensorLayout::NCX)
+                        .filter_layout(plaidml::op::TensorLayout::KCX)
+                        .autogroup_mode(plaidml::op::AutoGroupMode::EXPLICIT)
+                        .group_layout(plaidml::op::GroupLayout::IN_C)
+                        .group_layout(plaidml::op::GroupLayout::IN_K)
+                        .groups(static_cast<int>(G));
+      return edsl::make_tuple(result);
+    }
+    // Compute 2D DeformableConvolution.
+    if (rank == 4) {
+      // Define dim of each tensor.
+      auto H = I_shape[2], W = I_shape[3], OFF_H = OFF_shape[2], OFF_W = OFF_shape[3], F_H = F_shape[2],
+           F_W = F_shape[3];
+      // Throw exception
+      if (CI % G != 0 || CO % G != 0 || CI % DG != 0 || OFF_shape[1] % DG != 0) {
+        THROW_IE_EXCEPTION << "Incorrected shape for DeformableConvolution.";
+      }
+      // split offset along dg axis.
+      std::vector<edsl::Tensor> OFF_split_dgs;
+      for (auto i = 0; i < DG; ++i) {
+        edsl::Tensor OFF_split_dg = op::slice(OFF)
+                                        .add_dim(0, N)
+                                        .add_dim(i * OFF_C / DG, (i + 1) * OFF_C / DG)
+                                        .add_dim(0, OFF_H)
+                                        .add_dim(0, OFF_W);
+        OFF_split_dgs.push_back(OFF_split_dg);
+      }
+      // Get offsets of height and width.
+      std::vector<edsl::Tensor> offset_h_slices, offset_w_slices;
+      for (auto i = 0; i < DG; ++i) {
+        edsl::Tensor OFF_split_dg_h =
+            op::slice(OFF_split_dgs[i]).add_dim(0, N).add_dim(0, OFF_C / DG - 1, 2).add_dim(0, OFF_H).add_dim(0, OFF_W);
+        edsl::Tensor OFF_split_dg_w =
+            op::slice(OFF_split_dgs[i]).add_dim(0, N).add_dim(1, OFF_C / DG, 2).add_dim(0, OFF_H).add_dim(0, OFF_W);
+        edsl::Tensor offset_h_concat_axis_2(0), offset_w_concat_axis_2(0);
+        for (auto off_h = 0; off_h < OFF_H; ++off_h) {
```

Comment:
It seems to me what you are doing here in these two for loops can be replaced by transpose and reshape like the following. Please check and feel free to make changes. 
```
 edsl::Tensor single_offset_h = op::slice(OFF_split_dg_h)
                                     .add_dim(0, N)
                                     .add_dim(0, (OFF_C / DG) / 2)
                                     .add_dim(0, OFF_H)
                                     .add_dim(0, OFF_W);
single_offset_h = op::reshape(single_offset_h, make_tuple<int64_t>({N, 1, F_H, F_W, OFF_H, OFF_W}));

std::vector<Value> pattern;
pattern.push_back(Value{0});
pattern.push_back(Value{1});
pattern.push_back(Value{4});
pattern.push_back(Value{2});
pattern.push_back(Value{5});
pattern.push_back(Value{3});
single_offset_h = op::transpose(single_offset_h, edsl::Value{pattern});
single_offset_h = op::reshape(single_offset_h, make_tuple<int64_t>({N, 1, OFF_H*F_H,OFF_W*F_W }));

```